opm/grid/common/SubGridView.hpp

```diff
@@ -0,0 +1,371 @@
+
+
+    template <class BaseEntityType>
+    class SubEntity : public BaseEntityType
+    {
+    public:
+        SubEntity()
+            : BaseEntityType()
+            , is_owned_(false)
+        {
+        }
+        SubEntity(const BaseEntityType& base, const bool owned)
+            : BaseEntityType(base)
+            , is_owned_(owned)
+        {
+        }
+        auto partitionType() const
+        {
+            if (is_owned_) {
+                return Dune::InteriorEntity;
+            } else {
+                return Dune::OverlapEntity;
+            }
+        }
+    private:
+        bool is_owned_;
+    };
+
+    template <int cd>
+    struct Codim {
+        using BaseEntity = typename Grid ::Traits ::template Codim<cd>::Entity;
+        using Entity = SubEntity<BaseEntity>;
+        using EntitySeed = typename Grid ::Traits ::template Codim<cd>::EntitySeed;
+
+        using Geometry = typename Grid ::template Codim<cd>::Geometry;
+        using LocalGeometry = typename Grid ::template Codim<cd>::LocalGeometry;
+
+        /** \brief Define types needed to iterate over entities of a given partition type */
+        template <PartitionIteratorType pit>
+        struct Partition {
+            /** \brief iterator over a given codim and partition type */
+            using BaseIterator = typename Grid ::template Codim<cd>::template Partition<pit>::LeafIterator;
+        };
+    };
+
+    enum { conforming = Capabilities ::isLeafwiseConforming<Grid>::v };
+};
+
+
+template <class GridImp>
+class SubGridView
+{
+    using ThisType = SubGridView<GridImp>;
+
+public:
+    using Traits = SubGridViewTraits<GridImp>;
+
+    /** \brief type of the grid */
+    using Grid = typename Traits::Grid;
+
+    /** \brief type of the index set */
+    using IndexSet = typename Traits ::IndexSet;
+
+    /** \brief type of the intersection */
+    using Intersection = typename Traits ::Intersection;
+
+    /** \brief type of the intersection iterator */
+    using IntersectionIterator = typename Traits ::IntersectionIterator;
+
+    /** \brief type of the collective communication */
+    using CollectiveCommunication = typename Traits ::CollectiveCommunication;
+
+
+    /** \brief Codim Structure */
+    template <int cd>
+    struct Codim : public Traits ::template Codim<cd> {
+        using Entity = typename Traits::template Codim<cd>::Entity;
+        class SubIterator
+        {
+        public:
+            SubIterator(const SubGridView& view, std::size_t index)
+                : view_(&view)
+                , index_(index)
+            {
+            }
+            const Entity& operator*() const
+            {
+                entity_ = this->view_->get(index_);
+                return entity_;
+            }
+            const Entity* operator->() const
+            {
+                entity_ = this->view_->get(index_);
+                return &entity_;
+            }
```

Comment:
This seems very dangerous as another call the function will make the previously returned reference/pointer dangling.

Reply:
It will not be dangling, but it will suddenly refer to a different object. I agree this is a bit dangerous, but I did not see any way around the requirement for the operator. Suggestions are welcome. For the `operator*()` I guess we could return a copy instead, but that is not possible for `operator->()`.